hackathon_site/dashboard/frontend/src/components/user/LoginForm/loginform.test.js

```diff
@@ -87,6 +87,39 @@ describe("<EnhancedLoginForm />", () => {
         const button = getByText("Log In");
 
         expect(button.closest("button")).toBeDisabled();
-        expect(getByTestId("circular-progress")).toBeTruthy();
+        expect(getByTestId(TEST_IDS.circularProgress)).toBeTruthy();
+    });
+
+    it("Displays an alert when the rest failed for non-credentials reasons", () => {
+        const requestFailure = {
+            status: 500,
+            message: "Something went horribly wrong",
+        };
+        const { getByText, getByTestId } = render(
+            <EnhancedLoginForm
+                handleLogin={() => {}}
+                isLoading={false}
+                requestFailure={requestFailure}
+            />
+        );
+
+        expect(getByTestId(TEST_IDS.alert).innerHTML).toContain(requestFailure.message);
+    });
+
+    it("Displays field error states for invalid credentials", () => {
+        const requestFailure = {
+            status: 400,
+            message: "Invalid credentials",
+        };
+        const { queryByTestId, getAllByText } = render(
+            <EnhancedLoginForm
+                handleLogin={() => {}}
+                isLoading={false}
+                requestFailure={requestFailure}
+            />
+        );
+
+        expect(queryByTestId(TEST_IDS.alert)).toBeNull();
+        expect(getAllByText(ERROR_MESSAGES.credentialsInvalid).length).toEqual(2);
```

Comment:
You're right, I felt dirty about this test when I wrote it in the first place too. At the time I did this because I couldn't figure out how to get my way to the helper text from `getByText` on the inputs (which will actually return the input) - but I should just be able to go up to the parent and search for the text within. Will update :+1:

Reply:
Updated; technically the rest of the tests for error messages are actually imperfect since I don't test where they show up. The easiest way around that would be to snapshot test the form, but I don't really want to have all those snapshots (even though that's the point of them). Thoughts on this middle ground @rayraykay?

Reply:
Your most recent changes are just fine! And actually - I think I wasn't too clear on my initial message. I just meant that in the future anyone else working on such components and their unit tests should be wary of coverage and making sure that all possible conditions are covered (for ex. certain branches in their render like in this login form where an alert may pop up). As long as there's something easily queryable to get such an error message in a unit test and assert that it's there I don't think there's much more to add. Snapshots would most definitely be overkill - they should be testing that stuff renders with no special conditions and that's it